aten/src/ATen/native/AdaptiveAveragePooling.cpp

```diff
@@ -0,0 +1,330 @@
+        return AdaptiveAveragePooling2d_forward_out_cpu_template<scalar_t>(
+          output, input, output_size);
+      }
+    );
+    return output;
+  }
+
+  template <typename scalar_t>
+  Tensor AdaptiveAveragePooling2d_forward_cpu_template(
+    at::Tensor const& input,
+    IntList output_size
+  ) {
+    auto output = at::empty({0}, input.options());
+    AdaptiveAveragePooling2d_forward_out_cpu_template<scalar_t>(
+      output, input, output_size);
+    return output;
+  }
+
+  Tensor AdaptiveAveragePooling2d_forward_cpu(
+    at::Tensor const& input,
+    IntList output_size
+  ) {
+    return AT_DISPATCH_FLOATING_TYPES(input.type(), "adaptive_avg_pool2d", [&] {
+        return AdaptiveAveragePooling2d_forward_cpu_template<scalar_t>(
+          input, output_size);
+      });
+  }
+
+  template <typename scalar_t>
+  static void __AdaptiveAveragePooling2d_backward_out_frame(
+    scalar_t *gradInput_p,
+    scalar_t *gradOutput_p,
+    int64_t sizeD,
+    int64_t isizeH,
+    int64_t isizeW,
+    int64_t osizeH,
+    int64_t osizeW
+  ) {
+    int64_t d;
+  #pragma omp parallel for private(d)
+    for (d = 0; d < sizeD; d++)
+    {
+      scalar_t *gradInput_p_d = gradInput_p + d*isizeW*isizeH;
+      scalar_t *gradOutput_p_d = gradOutput_p + d*osizeW*osizeH;
+
+      /* calculate average */
+      int64_t oh, ow;
+      for(oh = 0; oh < osizeH; oh++)
+      {
+        int istartH = START_IND(oh, osizeH, isizeH);
+        int iendH   = END_IND(oh, osizeH, isizeH);
+        int kH = iendH - istartH;
+
+        for(ow = 0; ow < osizeW; ow++)
+        {
+
+          int istartW = START_IND(ow, osizeW, isizeW);
+          int iendW   = END_IND(ow, osizeW, isizeW);
+          int kW = iendW - istartW;
+
+          scalar_t grad_delta = gradOutput_p_d[oh*osizeW +ow] / kH / kW;
+
+          int ih, iw;
+          for(ih = istartH; ih < iendH; ih++)
+          {
+            for(iw = istartW; iw < iendW; iw++)
+            {
+              /* update gradient */
+              gradInput_p_d[ih*isizeW + iw] += grad_delta;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  template <typename scalar_t>
+  Tensor& AdaptiveAveragePooling2d_backward_out_cpu_template(
+    Tensor& gradInput,
+    const Tensor& gradOutput,
+    const Tensor& input
+  ) {
+    int dimD = 0;
+    int dimH = 1;
+    int dimW = 2;
+    int64_t sizeB = 1;
+    int sizeD;
+    int isizeH;
+    int isizeW;
+    int osizeH;
+    int osizeW;
+    scalar_t *gradInput_data;
+    scalar_t *gradOutput_data;
+
+    if (input.ndimension() == 4) {
+      sizeB = input.size(0);
+      dimD++;
+      dimH++;
+      dimW++;
+    }
+
+    /* sizes */
+    sizeD  = input.size(dimD);
+    isizeH = input.size(dimH);
+    isizeW = input.size(dimW);
+    osizeH = gradOutput.size(dimH);
+    osizeW = gradOutput.size(dimW);
+
+    /* get contiguous gradOutput */
+    auto gradOutput_ = gradOutput.contiguous();
```

Comment:
nit: I like to name the original with the underscore, and the contiguous version without, so that I don't accidentally access the wrong one later in the code